enhancements/cluster-logging/elasticsearch-hot-warm-architecture.md

```diff
@@ -0,0 +1,156 @@
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The purpose of this is to increase the potential retention of our logs and maintain that they can still be queried. Currently our logs are either maintained as a high priority index or they are deleted. We seek to emulate the back-end process that Elasticsearch makes available in the form of [x-pack API](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/index-lifecycle-management-api.html) end points.
+
+## Motivation
+
+Currently our managed Elasticsearch cluster is restricted in the amount of days of retention we can provide in part due to the amount of metadata ES keeps in memory related to its indices. If we are taking advantage of the different index management APIs provided by Elasticsearch we can seek to better tune clusters to increase retention and overall performance.
+
+### Goals
+The specific goals of this proposal are:
+
+* Outline the addition to the API to provide a means to configure WARM policies
+
+* Replace|repurpose the current rollover cronjob to also perform hot to warm transition actions
+
+We will be successful when:
+
+* Users are able to configure WARM policies so that we can optimize their cluster usage
+
+* We are able to transition a user's index from one policy to the next based on their configuration
+
+### Non-Goals
+
+* We will not be marking different nodes for the different policies and reallocating shards based on that
+
+* We will not be implementing this using the Opendistro plugin
+
+## Proposal
+
+Currently we offer a mechanism for users to configure their index lifecycles based on two states: hot and delete. By default indices are created and stay in the hot state, where they can be rolled over based on a configuration, however they stay in a "hot" state until they are eventually deleted based on the delete policy.
+
+Elasticsearch offers mechanisms to move indices between hot, warm, cold, and delete and discusses approaches for this [in a blog article](https://www.elastic.co/blog/implementing-hot-warm-cold-in-elasticsearch-with-index-lifecycle-management). However, this functionality is a x-pack feature and we do not support these. [Opendistro](https://opendistro.github.io/for-elasticsearch-docs/docs/ism/) offers a way to do this as well, however their plugins require ES7. We achieve our current functionality using cronjobs -- one for delete and one for rollover.
+
+Within the context of Elasticsearch ILM:
+1) Moving the index into a Warm state means the index is force merged to use less primary shards, its replica count is reduced, and the index priority is also reduced. The idea of this being that these indices are not being written to, may not be queried against as much so they do not need to be as distributed, and can be a lower priority when it comes to index recovery (meaning they would be recovered after high priority indices are).
```

Comment:
How does warm state translate in a number of primary shards and/or replica count? Do we want to be precise here?, e.g. warm = 2 primary shards, warm = 2 replicas?

Reply:
That is something we would likely play around with, but I would say lower than that. Maybe 1/1 or even 1/0.

Reply:
Looks like a case for examples to give third party readers a change to where we go.

Reply:
doesn't reducing the replica count compromise the redundancy policy?

Reply:
compromise it how? this is why users can specify when things go into "warm". The only thing that ends up "compromised" is in the case of recovery or how distributed data is with respect to being able to have multiple nodes fulfill the query request.

Reply:
Redundancy also speaks to the availability should an ES node never recover.  Reducing the replica count compromises that scenario

Reply:
I dont see it as compromising it as reframing it... the redundancy policy is for indices that are still "hot". I think it is important to note that something being rolled over isn't the same as it being "hot" or "warm" explicitly... also, it will be up to customers when they decide they want their indices to be moved into "warm" because they aren't as high a priority to be queried.

Most customers are looking for the grep experience which means after 24 hours they are less likely to look at these records, however they may still want to maintain logs for 7 days before deleting them. Without a warm policy, that's 6 days where indices aren't necessarily being queried and are taking up heap/disk space which competes with indices that are still "hot".